Python agent performance enhancement with asyncio (#316)
apache/skywalking-python: docs/en/setup/advanced/AsyncEnhancement.md

> # Python Agent Asynchronous Enhancement

Reviewer: This doc seems not to be linked from any place of the existing doc.
Do we need a new menu item?

Author: You mean I need to edit `docs/menu.yml` right?

It has been fixed.

## How it works

To keep the API unchanged, we have completely rewritten a new class called `SkyWalkingAgentAsync` (identical to the `SkyWalkingAgent` class). We use the environment variable mentioned above, `SW_AGENT_ASYNCIO_ENHANCEMENT`, to control which class implements the agent's interface.

In the `SkyWalkingAgentAsync` class, we have employed asyncio coroutines and their related functions to replace the Python threading implementation in nearly all instances. And we have applied asyncio enhancements to all three primary reporting protocols of the current SkyWalking Python agent:

- **gRPC**: We use the [`grpc.aio`](https://grpc.github.io/grpc/python/grpc_asyncio.html) module to replace the `grpc` module. Since the `grpc.aio` module is also officially supported and included in the `grpc` package, we can use it directly without any additional installation.

- **HTTP**: We use the [`aiohttp`](https://github.com/aio-libs/aiohttp) module to replace the `requests` module.

- **Kafka**: We use the [`aiokafka`](https://github.com/aio-libs/aiokafka) module to replace the `kafka-python` module.

> ## Proformance improvement

Reviewer 2: ```suggestion
## Performance improvement
```

> We using [wrk](https://github.com/wg/wrk) to pressure testing the network throughput of the Python agents in a [FastAPI](https://github.com/tiangolo/fastapi) application.

Reviewer 2: ```suggestion
We use [wrk](https://github.com/wg/wrk) to pressure test the network throughput of the Python agents in a [FastAPI](https://github.com/tiangolo/fastapi) application.
```

| :-------------: | :----: | :------: | :---------: |
| sync (original) | 345.89 | 56.41KB  |    1.09s    |
|   async (new)   | 655.67 | 106.93KB |    1.24s    |

> In fact, only the performance improvement of gRPC is of more reference value. Because the other two protocols use third-party libraries with completely different implementations, the performance improvement depends to a certain extent on the performance of these third-party libraries.

More details see this [PR](https://github.com/apache/skywalking-python/pull/316) .

## Potential problems

We have shown that the asynchronous enhancement function improves the transmission efficiency of metrics, traces and logs. But **it improves the proformance of profile data very little, and even causes performance degradation**.

This is mainly because a large part of the data in the `profile` part comes from the monitoring and measurement of Python threads, which is exactly what we need to avoid in asynchronous enhancement. Since operations on threads cannot be bypassed, we may need additional overhead to support cross-thread coroutine communication, which may lead to performance degradation instead of increase.

> Asynchronous enhancements involved many code changes and introduced some new dependencies. Since this feature is relatively new, it may cause some unexpected errors and problems. **If you encounter them, please feel free to contact us or submit issues and PRs**!

Reviewer 2: ```suggestion
Asynchronous enhancements involve many code changes and introduced some new dependencies. Since this feature is relatively new, it may cause some unexpected errors and problems. **If you encounter them, please feel free to contact us or submit issues and PRs**!
```
